src/ast/Expr.zig

```diff
@@ -647,13 +647,50 @@ pub fn jsonStringify(self: *const @This(), writer: anytype) !void {
     return try writer.write(Serializable{ .type = std.meta.activeTag(self.data), .object = "expr", .value = self.data, .loc = self.loc });
 }
 
+pub fn extractNumericValuesInSafeRange(left: Expr.Data, right: Expr.Data) ?[2]f64 {
+    const l_value = left.extractNumericValue() orelse return null;
+    const r_value = right.extractNumericValue() orelse return null;
+
+    // Check for NaN and return null if either value is NaN
+    if (std.math.isNan(l_value) or std.math.isNan(r_value)) {
+        return null;
+    }
+
+    if (std.math.isInf(l_value) or std.math.isInf(r_value)) {
+        return .{ l_value, r_value };
+    }
+
+    if (l_value > bun.jsc.MAX_SAFE_INTEGER or r_value > bun.jsc.MAX_SAFE_INTEGER) {
+        return null;
+    }
+    if (l_value < bun.jsc.MIN_SAFE_INTEGER or r_value < bun.jsc.MIN_SAFE_INTEGER) {
+        return null;
+    }
+
+    return .{ l_value, r_value };
+}
+
 pub fn extractNumericValues(left: Expr.Data, right: Expr.Data) ?[2]f64 {
     return .{
         left.extractNumericValue() orelse return null,
         right.extractNumericValue() orelse return null,
     };
 }
 
+pub fn extractStringValues(left: Expr.Data, right: Expr.Data, allocator: std.mem.Allocator) ?[2]*E.String {
+    const l_string = left.extractStringValue() orelse return null;
+    const r_string = right.extractStringValue() orelse return null;
+    l_string.resolveRopeIfNeeded(allocator);
+    r_string.resolveRopeIfNeeded(allocator);
+
+    if (l_string.isUTF8() != r_string.isUTF8()) return null;
```

Comment:
pretty sure when parsing JS E.String is either ascii or UTF16. we should allow comparing them because a conversion isn't necessary